### PR TITLE
Add minimal, interactive example using the DSL.

### PR DESCRIPTION
**Pull request by fwolfst**
My take on a "Hello World" example, based on `test/support/hello_world.rb`.

Several things to note:
  * I prefer the `bundle exec` style of development, but have seen commits "preventing" this. Please state the preferred style.
  * Without the `configure` (ll 18) block the examples do not work for me. Should I open an issue for that?
  * Directly rendering seems dirty to me. Without knowing the internals I wonder if this is the correct (or viable) way.
  * Merged like this, it would bring a examples folder into the gem. I do actually like it if installed gems ship with examples (currently test/support/examples is not included in the gem, see gemspec), but probably you want to wait for a cleaner example or have a better one at hand yourself.
  * I think the next example should feature a (in memory) persisted object to interact with (e.g. a counter :) ). In the given example I only see this achievable with databases or global variables, which are considered to be 80s.

### CODE REVIEW

File: examples/hello_worlds.rb [hunk 1]
```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/env ruby
+
+require 'vedeu'
```
Comment (gavinlaking) [inline examples/hello_worlds.rb]: I've started adding `require 'bundler/setup'` to the top of each of the executable examples. This means that you can use `bundle exec appname` or `./appname` freely. (I tested this with the `editor_app.rb` and it appeared to work as expected. Could you update your pull request to include this line, and let me know if you get the same behaviour.

```
require 'bundler/setup'
require 'vedeu'

fail Vedeu::VERSION
```

Try that in this example and let me know the outcome. If all is rosy, please add the `require 'bundler/setup'` line to this pull request.

File: examples/hello_worlds.rb [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+#
+# If you have cloned this repository from GitHub, you can run this example:
+#
+#     bundle exec ./examples/hello_worlds.rb
+#
+class HelloWorldsApp
+
+  Vedeu.bind(:_initialize_) { Vedeu.trigger(:_refresh_) }
+
+  Vedeu.configure do
```
Comment (gavinlaking) [inline examples/hello_worlds.rb]: Leave this in for now, but I have fixed this issue and will release a new version shortly.

File: examples/hello_worlds.rb [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  Vedeu.interface :hello do
+    # Define all of the interface in one place.
+    background '#000000'
+    foreground '#00ff00'
+    geometry do
+      centred!
+      height   5
+      width    24
+    end
+    Vedeu.views do
```
Comment (gavinlaking) [inline examples/hello_worlds.rb]: `Vedeu.views` are not really part of the interface DSL, but if this is working for you- then I can't see it as a problem!

File: examples/hello_worlds.rb [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+  def self.render_world
+    # Immediately render this to screen.
+    Vedeu.render do
+      view :hello do
+        geometry do
+          centred!
+          height  5 + 2
+          width   24 + 2
+        end
+        border :hello do
```
Comment (gavinlaking) [inline examples/hello_worlds.rb]: Although it doesn't hurt to be verbose, I don't think you need to name the border here as it will learn its name from `view :hello do` above.

**Comment by gavinlaking** [comment]
Let me know your thoughts on the above review points, and we'll get this merged in as soon as possible. :-D Nice work - thanks!

File: examples/hello_worlds.rb (hunk 1 above)
Comment (fwolfst) [inline examples/hello_worlds.rb]: As long as I do not have the development dependencies installed I can neither get it running with `bundle exec` nor `ruby example/hello_worlds.rb`, because `bundler/setup` is required somewhere within the vedeu/all, I suppose. If I uncomment, I get

    examples/hello_worlds.rb:5:in `<main>': 0.6.21 (RuntimeError)

File: examples/hello_worlds.rb (hunk 4 above)
Comment (fwolfst) [inline examples/hello_worlds.rb]: Right, I've seen that in the docs.

**Comment by fwolfst** [comment]
Was fun. Would be asome if you could comment on that:
 * Directly rendering seems dirty to me. Without knowing the internals I wonder if this is the correct (or viable) way.

And give me a keyword how to get forward about this:
 * I think the next example should feature a (in memory) persisted object to interact with (e.g. a counter :) ). In the given example I only see this achievable with databases or global variables, which are considered to be 80s. Attributes in the Application-class wont work, as there is no instance iiuc.

File: examples/hello_worlds.rb (hunk 1 above)
Comment (gavinlaking) [inline examples/hello_worlds.rb]: Sorry, I'm not sure if I've miscommunicated here. The `fail Vedeu::VERSION` line was simply to allow you to compare the versions of Vedeu.

I've just built a Docker image which will allow me to test in a 'bare-bones' fashion (who knows what crazy workarounds are in my development environment! :sweat_smile:.

Comment (fwolfst) [inline examples/hello_worlds.rb]: On 01.10.2015 17:10, Gavin Laking wrote:
> In examples/hello_worlds.rb
> <https://github.com/gavinlaking/vedeu/pull/244#discussion_r40924458>:
> 
>> @@ -0,0 +1,77 @@
>> +#!/usr/bin/env ruby
>> +
>> +require 'vedeu'
> 
> Sorry, I'm not sure if I've miscommunicated here. The |fail
> Vedeu::VERSION| line was simply to allow you to compare the versions of
> Vedeu.

Yes, but for my points that doesnt make a different (if it fails before
anything, it failed through the require `'bundler/setup'`.

> I've just built a Docker image which will allow me to test in a
> 'bare-bones' fashion (who knows what crazy workarounds are in my
> development environment! :sweat_smile:.

Well, you could also go with rvm and gemsets.
Have fun and thanks for looking into all that
--felix

**Comment by gavinlaking** [comment]
Regarding your last points, which I'll deal with in turn:

1) Directly rendering: When Vedeu first started out, this was the chosen way (by myself) to get content into a section of the screen (an interface). As Vedeu has developed, it has been made possible for you to create classes which contain the 'rendering' (`Vedeu.renders` / `Vedeu.views`) parts. Moving onwards, I then thought about how one could set up a view- an ERb template which contained both static content and dynamic content. This is available through:

```
Vedeu.renders do
  template_for(:your_interface ...
```

Take a look at `./test/support/examples/view_templates_app/view_templates_app.rb` for a sketchy example.

Once you have more than a couple of views you want to orchestrate with dynamic content, you start needing a framework- this is where I've tried to imitate an MVC style architecture with controllers, models and views all working together with Vedeu's help to produce a cohesive application. This of course produced the `vedeu_cli` gem.

Playa (https://github.com/gavinlaking/playa) is massively out of date in terms of Vedeu's development but gives you an idea of in memory persisted objects (the TrackCollection) being used to populate views. Sadly the code for Playa really doesn't give Vedeu the credit it deserves, as Vedeu is much tidier and more powerful than it was when I produced Playa.

Summing up, Vedeu needs more examples building and more documentation- something I am always trying to improve.

Thanks for your feedback so far, and if you have further questions, please don't hesitate. I'm going to merge this pull request shortly. :smile:

**Comment by fwolfst** [comment]
Thanks for the "historic" information. Would be great to have an example for each approach.

Still, I do not manage to build a simple interactive example, e.g. a uber-basic-file viewer:
 - main interface is split in two parts (east, west)
 - west shows list of files in current directory
 - east shows first 5 lines of a file if selected in west
 - you can navigate in west, the highlighted files first 5 lines are shown

I guess I struggle at two three points: 1) I do not understand menus. 2) I do not understand how (or why not) to bind (listen) to an existing event 3) I do not see where I can ask the interface where the cursor is located at (e.g. what content the line has).

I am sorry, have the feeling I miss some essential understanding.
